verticapy/stats/tools.py

```diff
@@ -1424,3 +1425,124 @@ def variance_inflation_factor(vdf: vDataFrame, X: list, X_idx: int = None):
+    """
+---------------------------------------------------------------------------
+Determines if there is a directed path from X -> Y given an adjacency list.
+
+Parameters
+----------
+adj_list: dict
+    Adjacency List where the values are the other keys
+X: str
+    Starting node
+Y: str
+    Ending Node
+
+Returns
+-------
+Boolean 
+True if there exists a path
+    """
+    if X not in adj_list.keys():
+        raise ParameterError(f"{X} not in Adjacency List")
+    if Y not in adj_list.keys():
+        raise ParameterError(f"{Y} not in Adjacency List")
+
+    visited = [X]
+    queue = [X]
+
+    while(queue):
+        m = queue.pop(0)
+        for neighbor in adj_list[m]:
+            if m not in adj_list[neighbor]: 
+                if neighbor == Y and m != X:
+                    return True
+                if neighbor not in visited:
+                    visited.append(neighbor)
+                    queue.append(neighbor)
+    return False
+    
+
+
+# ---#
+def conditional_chi_square(vdf: vDataFrame, X: str, Y: str, Z:list = [], alpha:float = None):
+    """
+---------------------------------------------------------------------------
+Computes the conditional chi square independence test. It can be used to detect
+spurious correlations.
+
+Parameters
+----------
+vdf: vDataFrame
+    Input vDataFrame.
+X: str
+    Input vColumn to Test.
+Y: str
+    Input vColumn to Test.
+    .
+Z: list
+    Conditional vColumns
+alpha: float
+    The probability the given distribution was generated randomly
+    Smaller alpha (ie close to 0) lead to denser graphs
+    Larger  alpha (ie close to 1) lead to sparser graphs
+
+Returns
+-------
+if alpha = None: 
+    (chi2: float, p_value: float, dof: int)
+if alpha != None: 
+    (p_value > alpha) : bool
+    """
+    if isinstance(Z, str):
+        Z = [Z]
+
+    check_types(
+        [("X", X, [str]), 
+         ("Y", Y, [str]),
+         ("Z", Z, [list]), 
+         ("alpha", alpha, [int, float]), 
+         ("vdf", vdf, [vDataFrame, str]),]
+    )
+    if isinstance(vdf, str):
+        vdf = vDataFrameSQL(relation=vdf)
+    
+    vdf.are_namecols_in(X) 
+    vdf.are_namecols_in(Y)   
+    vdf.are_namecols_in(Z)
+
+    relation = vdf.__genSQL__()
+    if Z == []:
+        table_sql = "SELECT {0}, {1}, count(*) as nij FROM {2} GROUP BY {0}, {1}".format(X, Y, relation)
+        contingency_table = vDataFrame(sql=table_sql)
+        crosstab_panda = contingency_table.to_pandas()
+        pivot = crosstab_panda.pivot(index=X, columns = Y).fillna(0)
+        chi, _, dof, _ = chi2_contingency(pivot)
+        pval = 1 - chi2.cdf(chi, df=dof)
+        if alpha == None:
+            return (chi, pval, dof)
+        return pval > alpha
+    
+    chi = 0
+    dof = 0
+    table_sql = "SELECT {0}, {1}, {2}, count(*) as nij FROM {3} GROUP BY {0}, {1}, {2}".format(', '.join(Z), X, Y, relation)
+    contingency_table = vDataFrame(sql=table_sql)
+    crosstab_panda = contingency_table.to_pandas()
```

Comment:
Same as above!

Reply:
To answer both as above. The reason why I chose to use pandas was that it was faster: it loaded the contingency_table into memory where I could do a lot of local calls instead of a bunch of system calls to Vertica. I originally had something below but with a very large dataset, constantly calling the **where** clause to create the `mini_table` was slower.

 

    unique_z = [["'{}'".format(x) for x in vdf[col].distinct()] for col in z]
    total_chi2 = 0
    total_dof = 0
    for condition in itertools.product(*unique_z):
        #SELECT a, d, c, b, count(*) AS "cnt" FROM square WHERE d=1 AND a=0 GROUP BY a , d, c , b
        sql_where_condition = " AND ".join(["{0}={1}".format(*x) for x in zip(z, condition)])
        mini_table_sql = "SELECT {0}, {1} FROM {3} WHERE {2};".format(x,y,sql_where_condition, vdf.__genSQL__())
        mini_table = vDataFrame(sql=mini_table_sql)
        pivot_table = mini_table.pivot_table_chi2(x)
        total_chi2 += pivot_table['chi2'][0]
        if pivot_table['chi2'][0] > 0:
            total_dof += pivot_table['dof'][0]
    return chi2.sf(total_chi2, total_dof) < alpha


I figured I could speed this up by first creating a contingency table in Vertica and then calling the **where** clause on that and then creating a pivot table.  A rough idea is what I have below. Except I was unable, over the summer, to figure out how to convert my **where** clause results into a proper pivot table.

    table_sql = "SELECT {0}, {1}, {2}, count(*) as nij FROM {3} GROUP BY {0}, {1}, {2}".format(', '.join(z), x, y, vdf.__genSQL__())
    contingency_table = vDataFrame(sql=table_sql)
    unique_z = [["'{}'".format(x) for x in vdf[col].distinct()] for col in z]
    
    total_chi2 = 0
    total_dof = 0
    for condition in itertools.product(*unique_z):
        #SELECT a, d, c, b, count(*) AS "cnt" FROM square WHERE d=1 AND a=0 GROUP BY a , d, c , b
        sql_where_condition = " AND ".join(["{0}={1}".format(*x) for x in zip(z, condition)])
        mini_table_sql = "SELECT {0}, {1}, nij FROM {3} WHERE {2};".format(x,y,sql_where_condition, contingency_table.__genSQL__())
        mini_table = vDataFrame(sql=mini_table_sql)
        display(mini_table.pivot_table([x,y]))
        pivot_table = mini_table.pivot_table_chi2(x)
        total_chi2 += pivot_table['chi2'][0]
        if pivot_table['chi2'][0] > 0:
            total_dof += pivot_table['dof'][0]
    return chi2.sf(total_chi2, total_dof) < alpha

So why doesn't this work?
1. `pivot_table_chi2()` expects all the data not a contingency_table
2. As shown in the files below, occasionally the pivot table (**expected result**) will have zeros so accounting for those locations given the `mini_table` proved to be difficult. Also below, I tried to run `pivot_table([x,y])` and I got those results.

Solutions:
1. Creating a quick temp table that is a contingency table so we don't need to rerun the same query to generate it because I believe currently to generate `mini_table` it runs `mini_table_sql` which reruns `contingency_table.__genSQL__()` which is time-consuming.
2. find a way to convert `mini_table` into **expected_result** within Vertica / VSQL commands. This would be almost `pivot_table` if it didn't aggregate data.

I think making this completely within Vertica is doable, but I don't think I have enough time to do it during this school semester.

files below
---

**mini_table**
![image](https://user-images.githubusercontent.com/25376671/200155136-d2f8e1af-4c72-4d85-847c-5d4aa450b6ee.png)

**mini_table.pivot_table([x,y])**
![image](https://user-images.githubusercontent.com/25376671/200155195-175c057b-82dd-49d3-81f2-3377855c08c8.png)

**expected result**
![image](https://user-images.githubusercontent.com/25376671/200155090-2d758e68-160d-4a3c-82d0-50f6ec8c5198.png)

Reply:
@zacandcheese can you please create an issue on that?